From 8872c86e1341571ffb4518f0d3a78cf43ec5affe Mon Sep 17 00:00:00 2001
From: "Cody J. Balos" <balos1@llnl.gov>
Date: Tue, 3 Nov 2020 11:43:58 -0800
Subject: [PATCH] dont use NULL in ltest for Hypre

---
 cmake/tpl/SundialsHypre.cmake | 7 +++----
 1 file changed, 3 insertions(+), 4 deletions(-)

diff --git a/cmake/tpl/SundialsHypre.cmake b/cmake/tpl/SundialsHypre.cmake
index 9633f1c..19d4bf2 100644
--- a/cmake/tpl/SundialsHypre.cmake
+++ b/cmake/tpl/SundialsHypre.cmake
@@ -78,15 +78,14 @@ if(HYPRE_FOUND AND (NOT HYPRE_WORKS))
   "ADD_EXECUTABLE(ltest ltest.c)\n"
   "TARGET_LINK_LIBRARIES(ltest ${HYPRE_LIBRARIES})\n")
 
-  # Create a C source file which calls a hypre function
   file(WRITE ${HYPRE_TEST_DIR}/ltest.c
   "\#include \"HYPRE_parcsr_ls.h\"\n"
   "int main(){\n"
   "HYPRE_ParVector par_b;\n"
   "HYPRE_IJVector b;\n"
-  "par_b = NULL;\n"
-  "b = NULL;\n"
-  "if (par_b != NULL || b != NULL) return(1);\n"
+  "par_b = 0;\n"
+  "b = 0;\n"
+  "if (par_b != 0 || b != 0) return(1);\n"
   "else return(0);\n"
   "}\n")
 
-- 
2.8.3

